docs/language-specification.md

```diff
@@ -0,0 +1,94 @@
+# p4-constraints language specification
+
+The general design philosophy behind the p4-constraints language is to recycle
+[P4 language] constructs as much as possible, both syntactically and
+semantically. Where the languages overlap, p4-constraints aims to be consistent
+with the [P4 language specification]. This is to minimize the cost of adaption
+for programmers already familiar with [P4].
+
+Where [P4] lacks constructs required in p4-constraints, we add new constructs
+guided by the principle of least surprise.
+
+[P4]: https://p4.org/specs/
+[P4 language]: https://p4.org/specs/
+[P4 language specification]: https://p4.org/specs/
+
+## Tour of the language
+
+The p4-constraints language is a simple, expression-based language for
+specifying properties of P4 table entries.
+
+### Keys and comparisons
+
+At the heart of the language lies the ability to talk about table keys,
+and to constrain their values using comparison operators.
+```p4
+@entry_restriction("ipv4.valid == 1")
+table ipv4_table { 
+  key {
+    ipv4.valid : exact;
+    ipv4.dst : lpm;
+  }
+}
+```
+For example, the above constraint `ipv4.valid == 1` states that the key
+`ipv4.valid` must be equal to `1` for all table entries. Just like P4, the
+constraint language provides the comparison operators `==`, `!=`, `>`, `>=`,
+`<`, and `<=` with the usual semantics.
+
+### Boolean operations
+
+The language provides the Boolean constants `true` and `false` for allowing
```

Comment:
I would reorder this to mention binary expressions first, that seems like the more useful part. The other is more of theoretical interest.

Reply:
Good point. I moved it to the end and got rid of the explanation what true and false might be good for (people are familiar with true and false).